### ISSUE TITLE
Flaky test: `TestTCPRouteEssentials`

### ISSUE THREAD

**Issue by pmalek** [Kong/kubernetes-ingress-controller#2668]
The following flaky test encountered during one of the PRs' tests:

```
time="2022-07-11T07:54:44Z" level=debug msg="fetching endpoints" service_name=tcpecho-1 service_namespace=045efcbc-725c-4aa9-8dad-f5b02b1adb39 service_port="&ServicePort{Name:tcp,Protocol:TCP,Port:8888,TargetPort:{0 1025 },NodePort:32354,AppProtocol:nil,}"
time="2022-07-11T07:54:44Z" level=debug msg="found endpoints: [{10.244.0.30 1025}]" service_name=tcpecho-1 service_namespace=045efcbc-725c-4aa9-8dad-f5b02b1adb39 service_port="&ServicePort{Name:tcp,Protocol:TCP,Port:8888,TargetPort:{0 1025 },NodePort:32354,AppProtocol:nil,}"
    tcproute_test.go:321: 
        	Error Trace:	/home/runner/work/kubernetes-ingress-controller/kubernetes-ingress-controller/test/integration/tcproute_test.go:321
        	Error:      	Condition never satisfied
        	Test:       	TestTCPRouteEssentials
    tcproute_test.go:169: cleaning up the tcproute 6c0ae890-38cc-449e-afa0-4554c1ca9f49
    tcproute_test.go:137: cleaning up the service tcpecho-1
    tcproute_test.go:99: cleaning up the deployments 045efcbc-725c-4aa9-8dad-f5b02b1adb39/tcpecho-1 and 045efcbc-725c-4aa9-8dad-f5b02b1adb39/tcpecho-2
    tcproute_test.go:40: unlocking TCP port
```

https://github.com/Kong/kubernetes-ingress-controller/runs/7278024888?check_suite_focus=true

**Comment by randmonkey** [comment]
one more case in nightly build: https://github.com/Kong/kubernetes-ingress-controller/runs/7462121679?check_suite_focus=true

**Comment by pmalek** [comment]
Recent occurrence: https://github.com/Kong/kubernetes-ingress-controller/runs/7747128878?check_suite_focus=true#step:5:957

**Comment by pmalek** [comment]
Still occurring: https://github.com/Kong/kubernetes-ingress-controller/runs/7763856269?check_suite_focus=true#step:6:1043

**Comment by rainest** [comment]
We don't appear to have much luck with consistent locations failing. On this:

```
2022-08-10T19:37:41.8402232Z time="2022-08-10T19:37:41Z" level=debug msg="parsing kubernetes objects into data-plane configuration"
2022-08-10T19:37:47.2784271Z     tcproute_test.go:193: 
2022-08-10T19:37:47.2786103Z         	Error Trace:	/home/runner/work/kubernetes-ingress-controller/kubernetes-ingress-controller/test/integration/tcproute_test.go:193
2022-08-10T19:37:47.2786994Z         	Error:      	Condition never satisfied
2022-08-10T19:37:47.2787962Z         	Test:       	TestTCPRouteEssentials
```

```
192     t.Log("verifying that the tcpecho is no longer responding")                                                               
193     require.Eventually(t, func() bool {                                                                                                                      
194         responded, err := tcpEchoResponds(fmt.Sprintf("%s:%d", proxyURL.Hostname(), ktfkong.DefaultTCPServicePort), testUUID1)
195         return responded == false && errors.Is(err, io.EOF)                                                               
196     }, ingressWait, waitTick)    
```
Diagnostics don't make the cause of that clear though, since config at the time was empty and the last pod logs indicate the route was indeed gone:

```
10.244.0.1 [10/Aug/2022:19:37:46 +0000] TCP 500 0 0 0.000                                                           
2022/08/10 19:37:46 [error] 2064#0: *8063 stream [lua] handler.lua:1200: before(): no Route found with those values while prereading client data, client: 10.244.0.1, server: 0.0.0.0:8888
```
I can't really think of why that wouldn't consistently return an EOF.

**Comment by pmalek** [comment]
Still failing: https://github.com/Kong/kubernetes-ingress-controller/runs/7880887179?check_suite_focus=true#step:5:723

**Comment by rainest** [comment]
Additional testing indicates the route was not responding, but that we got a reset instead of an EOF. Testing outside, it doesn't look like that's entirely unreasonable, as non-routed ports FIN the Kong side of the connection immediately and RST if you keep sending data:

```
14:19:05-0700 esenin $ nc -v localhost 9999
Connection to localhost (::1) 9999 port [tcp/distinct] succeeded!
a
a
a
```
```
2022/08/17 21:19:36 [error] 68597#0: *6201 stream [lua] handler.lua:1228: before(): no Route found with those values while prereading client data, client: 172.18.0.1, server: 0.0.0.0:9999
```
```
14:18:46.926583 lo    In  IP6 ::1.46046 > ::1.9999: Flags [S], seq 1085024688, win 65476, options [mss 65476,sackOK,TS val 2849626599 ecr 0,nop,wscale 7], length 0
14:18:46.926609 lo    In  IP6 ::1.9999 > ::1.46046: Flags [S.], seq 3320696775, ack 1085024689, win 65464, options [mss 65476,sackOK,TS val 2849626599 ecr 2849626599,nop,wscale 7], length 0
14:18:46.926632 lo    In  IP6 ::1.46046 > ::1.9999: Flags [.], ack 1, win 512, options [nop,nop,TS val 2849626599 ecr 2849626599], length 0
14:18:46.928616 lo    In  IP6 ::1.9999 > ::1.46046: Flags [F.], seq 1, ack 1, win 512, options [nop,nop,TS val 2849626601 ecr 2849626599], length 0
14:18:46.929699 lo    In  IP6 ::1.46046 > ::1.9999: Flags [.], ack 2, win 512, options [nop,nop,TS val 2849626602 ecr 2849626601], length 0
14:18:59.928241 lo    In  IP6 ::1.46046 > ::1.9999: Flags [P.], seq 1:3, ack 2, win 512, options [nop,nop,TS val 2849639601 ecr 2849626601], length 2
14:18:59.928282 lo    In  IP6 ::1.9999 > ::1.46046: Flags [.], ack 3, win 512, options [nop,nop,TS val 2849639601 ecr 2849639601], length 0
14:19:03.534700 lo    In  IP6 ::1.46046 > ::1.9999: Flags [P.], seq 3:5, ack 2, win 512, options [nop,nop,TS val 2849643207 ecr 2849639601], length 2
14:19:03.534740 lo    In  IP6 ::1.9999 > ::1.46046: Flags [.], ack 5, win 512, options [nop,nop,TS val 2849643207 ecr 2849643207], length 0
14:19:05.114948 lo    In  IP6 ::1.46046 > ::1.9999: Flags [P.], seq 5:7, ack 2, win 512, options [nop,nop,TS val 2849644788 ecr 2849643207], length 2
14:19:05.114970 lo    In  IP6 ::1.9999 > ::1.46046: Flags [R], seq 3320696777, win 0, length 0
```
It seems like we should still get an EOF on the read attempt, but it's possible that we're hitting the RST immediately after the write call in `tcpEchoResponds()`. Not sure why that would flake though--we send a constant-size message, so it seems like we should always get RST or always successfully write and get the EOF back after we attempt to read.

That said, not too concerned with whether there's a better way to check that we're indeed in the listening but no route state versus simply not listening: a RST could potentially mean that the port is simply closed, but that's quite unlikely given that we just successfully talked to it and will talk to it again. For the purposes of testing this, considering a RST equivalent should suffice.

Other TCP tests don't look affected, since they either don't check that routes disappear or are using a different probe that already fails on RSTs (e.g. they place an HTTP server behind it and check for a valid HTTP response, which a RST is not).

**Comment by rainest** [comment]
Closing this based on the lack of reports since the fix in #2834. If that wasn't sufficient, BuildPulse will raise the issue again.
